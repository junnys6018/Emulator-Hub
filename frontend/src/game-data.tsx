export type Console = 'NES' | 'GB' | 'GBC' | 'CHIP 8';

export interface GameMetaData {
    name: string;
    image: string;
    saveNames: string[];
    activeSaveIndex: number;
    console: Console;
    uuid: string;
}

interface Save {
    name: string;
    data: ArrayBuffer;
}

export interface GameData {
    rom: ArrayBuffer;
    saves: Save[];
    uuid: string;
}

import SMB from '@/public/assets/SMB.png';
export function useGameMetaData(): GameMetaData[] {
    return [
        {
            name: 'Super Mario Bros',
            image: SMB,
            saveNames: ['Save 1', 'Save 2', 'Save 3'],
            activeSaveIndex: 1,
            console: 'NES',
            uuid: '1',
        },
        {
            name: 'Zelda',
            image: SMB,
            saveNames: ['Save 1'],
            activeSaveIndex: 0,
            console: 'NES',
            uuid: '2',
        },
        {
            name: 'Donkey Kong',
            image: SMB,
            saveNames: ['Save 1'],
            activeSaveIndex: 0,
            console: 'NES',
            uuid: '3',
        },
        {
            name: 'Mario Cart',
            image: SMB,
            saveNames: ['Save 1'],
            activeSaveIndex: 0,
            console: 'NES',
            uuid: '4',
        },
        {
            name: 'Pokemon',
            image: SMB,
            saveNames: ['Save 1'],
            activeSaveIndex: 0,
            console: 'GB',
            uuid: '5',
        },
        {
            name: 'Super Mario World',
            image: SMB,
            saveNames: ['Save 1'],
            activeSaveIndex: 0,
            console: 'GBC',
            uuid: '6',
        },
        {
            name: 'Tetris',
            image: SMB,
            saveNames: ['Save 1'],
            activeSaveIndex: 0,
            console: 'GBC',
            uuid: '7',
        },
    ];
}
