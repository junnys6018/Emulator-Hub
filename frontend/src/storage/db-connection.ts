import { openDB } from 'idb';
import { EmulatorHubDB, initializeDatabase } from './storage';

const dbConnection = initializeDatabase().then(() => openDB<EmulatorHubDB>('emulator-hub'));

export default dbConnection;
