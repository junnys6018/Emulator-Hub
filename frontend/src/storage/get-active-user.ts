// TODO
/**
 * Returns the uuid of the currently active user
 * if no one is logged in, it is the uuid of the guest account
 * otherwise, it is the uuid of the logged in user
 *
 * Returns `null` if no active user can be found
 */
export default function getActiveUserUuid() {
    return localStorage.getItem('guest-uuid');
}
