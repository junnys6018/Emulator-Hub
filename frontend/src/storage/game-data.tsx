import { IDBPDatabase } from 'idb';
import _ from 'lodash';
import React, { useCallback, useContext, useEffect, useState } from 'react';
import getActiveUserUuid from './get-active-user';
import { EmulatorHubDB, Record, useDatabase } from './storage';
import { v4 as uuidv4 } from 'uuid';
import createSave from '../emulators/create-save';

export type Console = 'NES' | 'GB' | 'GBC' | 'CHIP 8';

export interface GameSettings {
    hidden: boolean;
    deletable: boolean;
    imageRendering: 'pixelated' | 'unset';
    // If true, a screenshot of the game will be taken next time it is played and will be used as the image for this rom
    captureImage: boolean;
}

export interface GameMetaData extends Record {
    name: string;
    image: Blob;
    saveNames: string[];
    activeSaveIndex: number;
    console: Console;
    user: string; // UUID of the user to which this record belongs to
    settings: GameSettings;
}

export interface GameMetaDataView {
    name: string;
    image: string;
    saveNames: string[];
    activeSaveIndex: number;
    console: Console;
    settings: GameSettings;
    uuid: string;
}

interface Save extends Record {
    data: ArrayBuffer | null;
}

export interface GameData extends Record {
    rom: ArrayBuffer;
    saves: Save[];
    user: string; // UUID of the user to which this record belongs to
}

const GameMetaDataContext = React.createContext<
    | [
          GameMetaDataView[],
          (newGameMetaData: RecursivePartial<GameMetaData>) => Promise<void>,
          (uuid: string) => Promise<void>,
      ]
    | null // Used to indicate gameMetaData is still loading
    | undefined // Default value, used to indicate context is being used without a provider
>(undefined);

function sortGameMetaData(g1: GameMetaDataView, g2: GameMetaDataView): number {
    const name1 = g1.name.toLowerCase();
    const name2 = g2.name.toLowerCase();

    if (name1 === name2) {
        return 0;
    } else if (name1 < name2) {
        return -1;
    }
    return 1;
}

export function GameMetaDataProvider(props: { children: React.ReactNode }) {
    const [gameMetaDataView, setGameMetaDataView] = useState<GameMetaDataView[] | null>(null);

    const db = useDatabase();

    useEffect(() => {
        const activeUuid = getActiveUserUuid();
        const transaction = db.transaction('gameMetaData');
        const objectStore = transaction.objectStore('gameMetaData');
        const userIndex = objectStore.index('by-user');

        userIndex.getAll(activeUuid).then(gameMetaData => {
            const gameMetaDataView: GameMetaDataView[] = [];
            for (const game of gameMetaData) {
                const image = URL.createObjectURL(game.image);
                gameMetaDataView.push({
                    name: game.name,
                    image: image,
                    saveNames: game.saveNames,
                    activeSaveIndex: game.activeSaveIndex,
                    console: game.console,
                    uuid: game.uuid,
                    settings: game.settings,
                });
            }

            gameMetaDataView.sort(sortGameMetaData);

            setGameMetaDataView(gameMetaDataView);
        });
    }, [db]);

    /**
     * Puts a `GameMetaData` record in the database
     *
     * If the record already exists, the new record is merged with the old one
     */
    const putGameMetaData = useCallback(
        async (newGameMetaData: Partial<GameMetaData> & { uuid: string }): Promise<void> => {
            if (gameMetaDataView === null) {
                return;
            }

            console.log('[INFO] updating/adding new gameMetaData');

            const existingImageUrl = gameMetaDataView.find(item => item.uuid === newGameMetaData.uuid)?.image;
            const overrideImage = existingImageUrl !== undefined && newGameMetaData.image !== undefined;
            if (overrideImage) {
                URL.revokeObjectURL(existingImageUrl);
            }

            const gameMetaData = await db.get('gameMetaData', newGameMetaData.uuid);
            const isNew = gameMetaData === undefined;
            if (!isNew) {
                const saveNames = newGameMetaData.saveNames;
                // Merge with existing item if it exists
                newGameMetaData = _.merge(gameMetaData, newGameMetaData);

                if (saveNames !== undefined) {
                    // If saveNames was specified, we dont want to merge the saves together, instead we overwrite the saveNames
                    newGameMetaData.saveNames = saveNames;
                }
            }

            // Create the new view
            const newGameMetaDataView = {
                name: newGameMetaData.name,
                image: overrideImage || isNew ? URL.createObjectURL(newGameMetaData.image as Blob) : existingImageUrl,
                saveNames: newGameMetaData.saveNames,
                activeSaveIndex: newGameMetaData.activeSaveIndex,
                console: newGameMetaData.console,
                uuid: newGameMetaData.uuid,
                settings: newGameMetaData.settings,
            } as GameMetaDataView;

            setGameMetaDataView(gameMetaDataView => {
                if (gameMetaDataView === null) {
                    return null;
                }

                if (isNew) {
                    return gameMetaDataView.concat(newGameMetaDataView).sort(sortGameMetaData);
                } else {
                    const uuid = newGameMetaData.uuid;

                    // Update the existing view, its important that we update the item instead of replacing it
                    // because code might save a reference to the view, in which case it will be holding onto
                    // a 'stale' reference of the view.
                    const existingGameMetaDataView = gameMetaDataView.find(
                        item => item.uuid === uuid,
                    ) as GameMetaDataView;
                    _.merge(existingGameMetaDataView, newGameMetaDataView);
                    existingGameMetaDataView.saveNames = newGameMetaDataView.saveNames;

                    // Create new array to force re-render
                    return [...gameMetaDataView].sort(sortGameMetaData);
                }
            });

            await db.put('gameMetaData', newGameMetaData as GameMetaData);
        },
        [db, gameMetaDataView],
    );

    const deleteGame = useCallback(
        async (uuid: string) => {
            // Delete the metadata
            setGameMetaDataView(gameMetaDataView => {
                if (gameMetaDataView === null) {
                    return null;
                }
                return gameMetaDataView.filter(value => value.uuid !== uuid);
            });

            // TODO: make this a single transaction
            await db.delete('gameMetaData', uuid);

            // Delete the game data
            await db.delete('gameData', uuid);
        },
        [db],
    );

    return (
        <GameMetaDataContext.Provider
            value={gameMetaDataView === null ? null : [gameMetaDataView, putGameMetaData, deleteGame]}
        >
            {gameMetaDataView !== null && props.children}
        </GameMetaDataContext.Provider>
    );
}

export function useGameMetaData() {
    const context = useContext(GameMetaDataContext);
    if (context === undefined) {
        throw new Error('useGameMetaData() must be called with a GameMetaDataProvider');
    } else if (context === null) {
        throw new Error('[BUG] useGameMetaData() called before data is loaded');
    }

    return context;
}

export function getGameData(db: IDBPDatabase<EmulatorHubDB>, uuid: string) {
    return db.get('gameData', uuid);
}

export async function putGameData(db: IDBPDatabase<EmulatorHubDB>, newGameData: Partial<GameData>) {
    const oldGameData = await db.get('gameData', newGameData.uuid as string);
    if (oldGameData !== undefined) {
        newGameData = _.merge(oldGameData, newGameData);
    }

    return db.put('gameData', newGameData as GameData);
}

export async function addSaveGame(db: IDBPDatabase<EmulatorHubDB>, uuid: string) {
    const gameData = await db.get('gameData', uuid);
    const gameMetaData = await db.get('gameMetaData', uuid);

    if (gameData === undefined || gameMetaData === undefined) {
        throw new Error('Invalid uuid');
    }

    gameData.saves.push({
        data: createSave(new Uint8Array(gameData.rom), gameMetaData.console),
        uuid: uuidv4(),
    });

    return db.put('gameData', gameData);
}

export async function deleteSaveGames(db: IDBPDatabase<EmulatorHubDB>, uuid: string, indices: Set<number>) {
    const gameData = await db.get('gameData', uuid);

    if (gameData === undefined) {
        throw new Error('Invalid uuid');
    }

    const newSaves = gameData.saves.filter((_, index) => !indices.has(index));
    gameData.saves = newSaves;

    return db.put('gameData', gameData);
}

export async function updateSave(db: IDBPDatabase<EmulatorHubDB>, uuid: string, index: number, save: ArrayBuffer) {
    const gameData = await db.get('gameData', uuid);
    if (gameData === undefined) {
        throw new Error('Invalid uuid');
    }

    gameData.saves[index].data = save;
    return db.put('gameData', gameData);
}
