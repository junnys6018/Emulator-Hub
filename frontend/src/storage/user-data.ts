import { Record } from './storage';
import { v4 as uuidv4 } from 'uuid';
import { useEffect, useState } from 'react';

export interface UserData extends Record {
    profileImage: Blob;
    userName: string;
}

interface UserProfile {
    profileImage: string;
    userName: string;
}

/**
 * Generate a new guest account with a random color gradient as the profile image
 *
 * May throw
 */
export async function generateGuestAccount(): Promise<UserData> {
    const canvas = document.createElement('canvas');
    canvas.width = 128;
    canvas.height = 128;

    const context = canvas.getContext('2d');
    if (!context) {
        throw new Error('Failed to create canvas context');
    }

    const getRandomColor = () => {
        return `hsl(${Math.random() * 256}, 100%, 50%)`;
    };

    const gradient = context.createLinearGradient(0, 0, 128, 128);
    gradient.addColorStop(0, getRandomColor());
    gradient.addColorStop(1, getRandomColor());

    context.fillStyle = gradient;
    context.fillRect(0, 0, 128, 128);

    const profileImage = await new Promise<Blob>((resolve, reject) => {
        canvas.toBlob(blob => {
            if (blob) {
                resolve(blob);
            } else {
                reject(new Error('Failed to serialize image to a blob'));
            }
        }, 'image/png');
    });

    return {
        uuid: uuidv4(),
        age: 0,
        userName: 'Guest',
        profileImage: profileImage,
    };
}

const whiteImage =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACABAMAAAAxEHz4AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAADUExURf///6fEG8gAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAfSURBVGje7cExAQAAAMKg9U9tCF8gAAAAAAAAAIBLDSCAAAEf1udwAAAAAElFTkSuQmCC';
export function useUserProfile() {
    const [userProfile, setUserProfile] = useState<UserProfile>({
        profileImage: whiteImage,
        userName: 'Loading...',
    });

    useEffect(() => {
        // setup
        let url = '';
        generateGuestAccount()
            .then(userData => {
                url = URL.createObjectURL(userData.profileImage);
                setUserProfile({
                    profileImage: url,
                    userName: userData.userName,
                });
            })
            .catch(error => {
                alert(error);
            });

        // cleanup
        return () => {
            if (url) {
                URL.revokeObjectURL(url);
            }
        };
    }, []);

    return userProfile;
}
