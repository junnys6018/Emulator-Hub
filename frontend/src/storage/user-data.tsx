import { Record, EmulatorHubDB } from './storage';
import { v4 as uuidv4 } from 'uuid';
import React, { useContext, useEffect, useState } from 'react';
import { openDB } from 'idb';

export interface UserData extends Record {
    profileImage: Blob;
    userName: string;
}

interface UserProfile {
    profileImage: string;
    userName: string;
}

/**
 * Generate a new guest account with a random color gradient as the profile image
 *
 * May throw
 */
export async function generateGuestAccount(): Promise<UserData> {
    const canvas = document.createElement('canvas');
    canvas.width = 128;
    canvas.height = 128;

    const context = canvas.getContext('2d');
    if (!context) {
        throw new Error('Failed to create canvas context');
    }

    const getRandomColor = () => {
        return `hsl(${Math.random() * 256}, 100%, 50%)`;
    };

    const gradient = context.createLinearGradient(0, 0, 128, 128);
    gradient.addColorStop(0, getRandomColor());
    gradient.addColorStop(1, getRandomColor());

    context.fillStyle = gradient;
    context.fillRect(0, 0, 128, 128);

    const profileImage = await new Promise<Blob>((resolve, reject) => {
        canvas.toBlob(blob => {
            if (blob) {
                resolve(blob);
            } else {
                reject(new Error('Failed to serialize image to a blob'));
            }
        }, 'image/png');
    });

    return {
        uuid: uuidv4(),
        age: 0,
        userName: 'Guest',
        profileImage: profileImage,
    };
}

const whiteImage =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAIAAAACABAMAAAAxEHz4AAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAADUExURf///6fEG8gAAAAJcEhZcwAADsMAAA7DAcdvqGQAAAAfSURBVGje7cExAQAAAMKg9U9tCF8gAAAAAAAAAIBLDSCAAAEf1udwAAAAAElFTkSuQmCC';

const UserProfileContext = React.createContext<UserProfile>({
    profileImage: whiteImage,
    userName: 'Loading...',
});

export function UserProfileProvider(props: { children: React.ReactNode }) {
    const [userProfile, setUserProfile] = useState<UserProfile>({
        profileImage: whiteImage,
        userName: 'Loading...',
    });

    useEffect(() => {
        let url = '';
        const guestUuid = localStorage.getItem('guest-uuid');

        const loadUserDataFromUuid = (uuid: string) => {
            openDB<EmulatorHubDB>('emulator-hub', 1)
                .then(db => db.get('users', uuid))
                .then(userData => {
                    if (userData) {
                        url = URL.createObjectURL(userData.profileImage);
                        setUserProfile({
                            profileImage: url,
                            userName: userData.userName,
                        });
                    } else {
                        // TODO
                        alert(`[FATAL] uuid ${uuid} does not point to an account`);
                    }
                });
        };

        if (guestUuid) {
            console.log('[INFO] found guestUuid, loading from storage');
            loadUserDataFromUuid(guestUuid);
        } else {
            console.log('[INFO] guestUuid not found, adding event listener...');
            window.addEventListener('storage', function callback(e) {
                if (e.key === 'guest-uuid' && e.newValue !== null) {
                    console.log('[INFO] guestUuid found, removing event listener...');
                    loadUserDataFromUuid(e.newValue);
                    window.removeEventListener('storage', callback);
                }
            });
        }

        return () => {
            if (url) {
                URL.revokeObjectURL(url);
            }
        };
    }, []);

    return <UserProfileContext.Provider value={userProfile}>{props.children}</UserProfileContext.Provider>;
}

export function useUserProfile() {
    return useContext(UserProfileContext);
}
