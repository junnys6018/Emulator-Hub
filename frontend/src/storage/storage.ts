import { openDB, DBSchema, IDBPDatabase } from 'idb';
import { generateGuestAccount, UserData } from './user-data';
import { v4 as uuidv4 } from 'uuid';

export interface Record {
    uuid: string;
    age: number;
}

export interface EmulatorHubDB extends DBSchema {
    users: {
        value: UserData;
        key: string;
    };
}

export async function createGuestAccount(db: IDBPDatabase<EmulatorHubDB>) {
    console.log('[INFO] No guest account found, generating one');
    const guestUser = await generateGuestAccount();

    const addGuestAccount = async () => {
        try {
            await db.add('users', guestUser);
        } catch (error) {
            if (error.name === 'ConstraintError') {
                console.warn('[WARN] primary key collision, regenerating uuid');
                guestUser.uuid = uuidv4();
                await addGuestAccount();
            } else {
                throw error;
            }
        }
    };
    await addGuestAccount();

    localStorage.setItem('guest-uuid', guestUser.uuid);
    // Dispatch a storage event, this event is not triggered on the window object that modified localStorage
    // so we do it manually
    const event = new StorageEvent('storage', {
        key: 'guest-uuid',
        newValue: guestUser.uuid,
    });
    window.dispatchEvent(event);
}

export async function initializeDatabase() {
    const db = await openDB<EmulatorHubDB>('emulator-hub', 1, {
        upgrade(db, oldVersion) {
            switch (oldVersion) {
                case 0:
                    // database didnt exist, initialize schema
                    console.log('[INFO] Upgrading database to version 1');
                    db.createObjectStore('users', { keyPath: 'uuid' });
            }
        },
        //blocked() {},
        blocking() {
            db.close();
            alert('The application is outdated, please reload the page');
        },
        terminated() {
            alert('[FATAL] database connection was unexpectedly closed');
        },
    });

    // Check if a guest account exists, if not create one
    const guestUuid = localStorage.getItem('guest-uuid');
    if (!guestUuid) {
        await createGuestAccount(db);
    }

    db.close();
}
