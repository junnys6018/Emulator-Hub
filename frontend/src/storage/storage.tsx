import { openDB, DBSchema, IDBPDatabase, OpenDBCallbacks } from 'idb';
import { UserData } from './user-data';
import React, { useContext, useEffect, useState } from 'react';
import { useAlert } from '../components/util/alert';
import { GameData, GameMetaData } from './game-data';
import getActiveUserUuid from './get-active-user';

export interface Record {
    uuid: string;
}

export interface EmulatorHubDB extends DBSchema {
    users: {
        value: UserData;
        key: string;
    };
    gameMetaData: {
        value: GameMetaData;
        key: string;
        indexes: { 'by-user': string };
    };
    gameData: {
        value: GameData;
        key: string;
        indexes: { 'by-user': string };
    };
}

const DatabaseContext = React.createContext<IDBPDatabase<EmulatorHubDB> | null>(null);

export async function initializeDatabase(name: string, options?: OpenDBCallbacks<EmulatorHubDB>) {
    const db = await openDB<EmulatorHubDB>(name, 1, {
        upgrade(db, oldVersion) {
            switch (oldVersion) {
                case 0:
                    // database didnt exist, initialize schema
                    console.log('[INFO] Upgrading database to version 1');
                    db.createObjectStore('users', { keyPath: 'uuid' });

                    const gameMetaData = db.createObjectStore('gameMetaData', { keyPath: 'uuid' });
                    gameMetaData.createIndex('by-user', 'user');

                    const gameData = db.createObjectStore('gameData', { keyPath: 'uuid' });
                    gameData.createIndex('by-user', 'user');
            }
        },
        blocked: options?.blocked,
        blocking() {
            db.close();
            if (options?.blocking) options.blocking();
        },
        terminated: options?.terminated,
    });

    return db;
}

export function DatabaseProvider(props: { children?: React.ReactNode; name?: string }) {
    const [db, setDb] = useState<IDBPDatabase<EmulatorHubDB> | null>(null);
    const alert = useAlert();

    const name = props.name ?? 'emulator-hub';

    useEffect(() => {
        initializeDatabase(name, {
            blocking() {
                alert('The application is outdated, please reload the page', {
                    action: 'REFRESH',
                });
            },
            terminated() {
                alert('The database connection was unexpectedly closed', {
                    severity: 'ERROR',
                    action: 'REFRESH',
                });
            },
        }).then(async db => {
            // Health Check
            const activeUuid = getActiveUserUuid();
            if (activeUuid !== null) {
                const user = await db.get('users', activeUuid);
                if (user === undefined) {
                    // active-uuid points to non-existent user
                    console.log('[WARN] active-uuid did not point to a valid user');
                    localStorage.removeItem('active-uuid');
                }
            }

            setDb(db);
        });
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return <DatabaseContext.Provider value={db}>{db !== null && props.children}</DatabaseContext.Provider>;
}

export function useDatabase() {
    const db = useContext(DatabaseContext);
    if (db === null) {
        throw new Error('useDatabase must be used within DatabaseProvider');
    }
    return db;
}
