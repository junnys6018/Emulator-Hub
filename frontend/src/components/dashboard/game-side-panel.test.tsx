/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import renderer from 'react-test-renderer';
import GameSidePanel, { newSaveIndex } from './game-side-panel';
import { useGameMetaData } from '@/src/storage/game-data';

jest.mock('@/src/storage/game-data');
jest.mock('../util/alert');

test('<GameSidePanel /> snapshot', () => {
    const props = {
        image: '/image.png',
        imageRendering: 'pixelated' as const,
        name: 'Zelda',
        gameUuid: 'mock',
        saveNames: ['Save 1', 'Save 2', 'Save 3'],
        activeSaveIndex: 1,
        closePanel: () => {},
    };

    (useGameMetaData as jest.Mock<any, any>).mockReturnValueOnce([, jest.fn()]);

    const tree = renderer.create(<GameSidePanel {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test('newSaveIndex', () => {
    expect(newSaveIndex(new Set([]), 3)).toBe(3);
    expect(newSaveIndex(new Set([4, 5, 6]), 3)).toBe(3);
    expect(newSaveIndex(new Set([1, 4, 5]), 3)).toBe(2);
    expect(newSaveIndex(new Set([3]), 3)).toBe(0);
    expect(newSaveIndex(new Set([1, 2, 6]), 3)).toBe(1);
    expect(newSaveIndex(new Set([0, 1, 2, 6]), 3)).toBe(0);
    expect(newSaveIndex(new Set([0]), 3)).toBe(2);
});
