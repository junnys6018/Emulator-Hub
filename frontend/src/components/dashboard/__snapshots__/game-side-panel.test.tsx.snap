// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameSidePanel /> snapshot 1`] = `
Array [
  <div
    className="h-1 flex-shrink-0 bg-primary-500"
  />,
  <div
    className="flex flex-grow flex-col overflow-y-auto bg-gray-800"
  >
    <div
      className="container flex flex-col"
    >
      <div
        className="flex items-center py-5 md:py-8"
      >
        <button
          className="mr-3 flex-shrink-0 active:text-green-500 md:hover:text-green-500"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height="20px"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
            />
          </svg>
        </button>
        <h1
          className="mr-auto font-semibold text-2xl truncate"
        >
          Zelda
        </h1>
        <button
          className="flex-shrink-0 active:text-green-500 md:hover:text-green-500"
        >
          <svg
            fill="currentColor"
            height="20px"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 576 512"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
            />
          </svg>
        </button>
      </div>
      <img
        className="rounded-2xl pixelated drop-shadow object-center object-cover w-full md:w-96 mx-auto"
        src="/image.png"
        style={
          Object {
            "aspectRatio": "1",
          }
        }
      />
      <h2
        className="font-medium text-2xl pt-6 pb-2 md:py-6"
      >
        Saves
      </h2>
    </div>
    <div
      className="
                    bg-gray-700
                    
                "
    >
      <div
        className="container h-12 flex items-center"
      >
        <span
          className="text-lg mr-auto "
        >
          Save 1
        </span>
      </div>
    </div>
    <div
      className="
                    bg-gray-800
                    text-primary-500
                "
    >
      <div
        className="container h-12 flex items-center"
      >
        <span
          className="text-lg mr-auto game-side-panel__circle"
        >
          Save 2
        </span>
        <span
          className="text-lg"
        >
          Active
        </span>
      </div>
    </div>
    <div
      className="
                    bg-gray-700
                    
                "
    >
      <div
        className="container h-12 flex items-center"
      >
        <span
          className="text-lg mr-auto "
        >
          Save 3
        </span>
      </div>
    </div>
    <div
      className="container flex flex-col mt-2"
    >
      <button
        className="text-gray-300 active:text-green-500 md:hover:text-green-500 flex items-center mb-9 w-max"
      >
        <span
          className="font-medium text-xs mr-1"
        >
          Add Save
        </span>
        <svg
          fill="currentColor"
          height="8px"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="8px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
      </button>
      <h2
        className="font-medium text-2xl mb-2"
      >
        Settings
      </h2>
      <form
        className="mb-10"
      >
        <input
          className="mr-2"
          id="hidden"
          name="hidden"
          type="checkbox"
        />
        <label
          className="text-sm font-medium"
          htmlFor="hidden"
        >
          Hidden
        </label>
      </form>
    </div>
    <div
      className="container mt-auto"
    >
      <button
        className="btn-primary mb-10 h-12 w-full"
      >
        <svg
          className="mr-4"
          fill="currentColor"
          height="12px"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="12px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M424.4 214.7L72.4 6.6C43.8-10.3 0 6.1 0 47.9V464c0 37.5 40.7 60.1 72.4 41.3l352-208c31.4-18.5 31.5-64.1 0-82.6z"
          />
        </svg>
        <span
          className="font-medium text-2xl"
        >
          Play
        </span>
      </button>
    </div>
  </div>,
]
`;
